mysql/mysql/mysql.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package mysql provides a Cloud SQL MySQL driver that uses go-sql-driver/mysql
+// and works with database/sql
+package mysql
+
+import (
+	"context"
+	"database/sql"
+	"database/sql/driver"
+	"net"
+
+	"cloud.google.com/go/cloudsqlconn"
+	"github.com/go-sql-driver/mysql"
+)
+
+// RegisterDriver registers a MySQL driver that uses the cloudsqlconn.Dialer
+// configured with the provided options. The choice of name is entirely up to
+// the caller and may be used to distinguish between multiple registrations of
+// differently configured Dialers.
+func RegisterDriver(name string, opts ...cloudsqlconn.Option) (func() error, error) {
+	sql.Register(name, &mysqlDriver{})
+	d, err := cloudsqlconn.NewDialer(context.Background(), opts...)
+	if err != nil {
+		return func() error { return nil }, err
+	}
+	mysql.RegisterDialContext(name, mysql.DialContextFunc(func(ctx context.Context, addr string) (net.Conn, error) {
+		return d.Dial(ctx, addr)
+	}))
+	return func() error {
+		d.Close()
+		return nil
+	}, nil
+}
+
+type dialerConn struct {
+	driver.Conn
+	dialer *cloudsqlconn.Dialer
+}
+
+func (c *dialerConn) Close() error {
+	c.dialer.Close()
+	return c.Conn.Close()
+}
+
+type mysqlDriver struct {
+	d *cloudsqlconn.Dialer
+}
+
+// Open accepts a DSN using the go-sql-driver/mysql format. See
+// https://github.com/go-sql-driver/mysql#dsn-data-source-name for details.
+// Note the protocol should match the name used when registering a driver. For
+// example, a connection string might look like this where "cloudsql-mysql" is
+// the named used when registering the driver:
+//
+//     my-user:mypass@cloudsql-mysql(my-proj:us-central1:my-inst)/my-db
+//
+func (d *mysqlDriver) Open(name string) (driver.Conn, error) {
+	drv := &mysql.MySQLDriver{}
```

Comment:
We manually initialize a driver here to mimic what happens when a user imports the MySQL package (see [here](https://github.com/GoogleCloudPlatform/cloud-sql-go-connector/pull/80)).

Reply:
This allows us to encapsulate the MySQL driver under whatever name was registered above.

Reply:
IMO, it's better to initialize the driver once rather than create a new one every time we need to call open. 

I would probably move it into the mysqlDriver struct, so it's created when the driver is registered.

Reply:
Agreed.